specs/emip-005.md

```diff
@@ -0,0 +1,472 @@
+
+Cardano uses BECH32 in what it called `v3` addresses defined in [CIP-001](https://github.com/input-output-hk/implementation-decisions/blob/master/text/0001-address.md)
+
+Although `v3` is relevant to us for encoding public keys into addresses, we do not use `bech32`'s scheme for encoding in this specification. This is because
+
+- The payload may be too big to reasonably encode in a QR code so the benefit of using base32 are limited.
+- BCH checksums are not made for large payloads and additionally the polynomial used has to be fine-tuned for the expected length (but the length of our payload varies too much in this spec)
+
+### CRC32
+
+Although there is no specification for CRC32 and many variations exist, in Cardano we use the CRC-32-IEEE variation. You can find a C implementation [here](https://github.com/cardano-foundation/ledger-app-cardano/blob/3f784d23c1b87df73cda552ef01428d3e2733237/src/crc32.c#L6)
+
+### Blake2b [[RFC 7693](https://tools.ietf.org/html/rfc7693)]
+
+Blake2b is a hash algorithm used commonly in Cardano. Notably, Blake2b-224, Blake2b-256 and Blake2b512 are used depending on the context.
+
+### base64url [[RFC 4648 section 5](https://tools.ietf.org/html/rfc4648#section-5)]
+
+`base64url` allows to encode bytes in a human-readable format that is also safe to pass in URLs.
+
+### Other blockchain standards
+
+Other blockchains have existing specifications for message signing, but they mostly revolve around scripts trying to validate messages. We don't leverage any of their work in particular but it may be of interest.
+
+- [BIP-137](https://github.com/bitcoin/bips/blob/master/bip-0137.mediawiki) - simply scheme for message signing that works with P2PKH, P2PSH and bech32
+- [BIP-322](https://github.com/kallewoof/bips/blob/master/bip-0322.mediawiki) - reuses Bitcoin script to process a generic signed message format
+
+- [EIP-191](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-191.md) - encode data for Ethereum contracts
+- [EIP-712](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md) - encode structs for Ethereum contracts
+
+# Existing Code
+
+### Cryptography
+
+Cardano already allows message signing within the WASM bindings. Notably,
+
+1) [sign](https://github.com/input-output-hk/js-cardano-wasm/blob/master/cardano-wallet/src/lib.rs#L177)
+2) [verify](https://github.com/input-output-hk/js-cardano-wasm/blob/master/cardano-wallet/src/lib.rs#L232)
+
+You can see an example of these two functions [here](https://repl.it/repls/GlossyDarkgreenMethod)
+
+Even if you use cryptographically sound `sign` and `verify` functions, you still have the following problems:
+
+1) No human-recognizable prefix
+0) No error detection
+0) User could accidentally sign a transaction or a block thinking it's harmless data
+
+We also have a risk of a few different kinds of replay attacks
+
+4) A dapp asks person A to sign "BOB" and then another dapp asks user B to sign "BOB". B can just use the signature from A
```

Comment:
What is a dapp?

Reply:
A dapp is a "decentralized application". Usually generically refers to any application that uses a smart contract to power its core functionality.